File: run/src/main/java/org/screamingsandals/gradle/run/RunTestServerExtension.java
/*
 * Copyright 2024 ScreamingSandals
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.screamingsandals.gradle.run;

import lombok.AccessLevel;
import lombok.Data;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import org.gradle.api.Action;
import org.gradle.api.file.RegularFile;
import org.gradle.api.provider.Provider;
import org.gradle.api.provider.ProviderFactory;
import org.jetbrains.annotations.ApiStatus;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.screamingsandals.gradle.run.config.MultipleVersions;
import org.screamingsandals.gradle.run.config.Platform;
import org.screamingsandals.gradle.run.config.Version;

import javax.inject.Inject;
import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

@Data
@RequiredArgsConstructor(onConstructor_ = {@Inject})
public class RunTestServerExtension {
    @Getter(AccessLevel.NONE)
    private final @NotNull ProviderFactory providers;
    private @NotNull List<@NotNull Version> versions = new ArrayList<>();
    private @Nullable Provider<RegularFile> pluginJar;
    private @NotNull String testingDirectory = "test-environment";

    public @NotNull Version version(@NotNull Version version) {
        this.versions.add(version);
        return version;
    }

    public @NotNull Version version(@NotNull Version version, @NotNull Action<@NotNull Version> callback) {
        this.versions.add(version);
        callback.execute(version);
        return version;
    }

    public @NotNull Version version(@NotNull Platform platform, @NotNull String version) {
        return version(new Version(platform, version, platform.name().toLowerCase(Locale.ROOT) + "/" + version));
    }

    public @NotNull Version version(@NotNull Platform platform, @NotNull String version, @NotNull Action<@NotNull Version> callback) {
        var server = version(platform, version);
        callback.execute(server);
        return server;
    }

    public @NotNull MultipleVersions versions(@NotNull Platform platform, @NotNull String @NotNull... versions) {
        var list = new ArrayList<Version>();
        for (var version : versions) {
            list.add(version(platform, version));
        }
        return new MultipleVersions(list);
    }

    public @NotNull MultipleVersions versions(@NotNull Platform platform, @NotNull List<@NotNull String> versions) {
        var list = new ArrayList<Version>();
        for (var version : versions) {
            list.add(version(platform, version));
        }
        return new MultipleVersions(list);
    }

    public @NotNull MultipleVersions versions(@NotNull Platform platform, @NotNull List<@NotNull String> versions, @NotNull Action<@NotNull MultipleVersions> callback) {
        var multipleVersions = versions(platform, versions);
        callback.execute(multipleVersions);
        return multipleVersions;
    }

    public @NotNull Version paper(@NotNull String version) {
        return version(Platform.PAPER, version);
    }

    public @NotNull Version paper(@NotNull String version, @NotNull Action<@NotNull Version> callback) {
        return version(Platform.PAPER, version, callback);
    }

    public @NotNull MultipleVersions paperVersions(@NotNull String @NotNull... versions) {
        return versions(Platform.PAPER, versions);
    }

    public @NotNull MultipleVersions paperVersions(@NotNull List<@NotNull String> versions) {
        return versions(Platform.PAPER, versions);
    }

    public @NotNull MultipleVersions paperVersions(@NotNull List<@NotNull String> versions, @NotNull Action<@NotNull MultipleVersions> callback) {
        return versions(Platform.PAPER, versions, callback);
    }

    public void pluginJar(@NotNull Provider<RegularFile> pluginJar) {
        this.pluginJar = pluginJar;
    }

    @ApiStatus.Obsolete
    public void pluginJar(@NotNull File pluginJar) {
        this.pluginJar = providers.provider(() -> () -> pluginJar);
    }

    public void testingDirectory(@NotNull String testingDirectory) {
        this.testingDirectory = testingDirectory;
    }
}

File: run/src/main/kotlin/org/screamingsandals/gradle/run/RunTestServerExtension.kt
/*
 * Copyright 2025 ScreamingSandals
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.screamingsandals.gradle.run

import org.gradle.api.Action
import org.gradle.api.file.RegularFile
import org.gradle.api.provider.Provider
import org.gradle.api.provider.ProviderFactory
import org.jetbrains.annotations.ApiStatus
import org.screamingsandals.gradle.run.config.MultipleVersions
import org.screamingsandals.gradle.run.config.Platform
import org.screamingsandals.gradle.run.config.Version
import java.io.File
import javax.inject.Inject

open class RunTestServerExtension @Inject constructor(private val providers: ProviderFactory) {
    var versions: MutableList<Version> = mutableListOf()
    var pluginJar: Provider<RegularFile>? = null
    var testingDirectory: String = "test-environment"

    fun version(version: Version): Version {
        this.versions.add(version)
        return version
    }

    fun version(version: Version, callback: Action<Version>): Version {
        this.versions.add(version)
        callback.execute(version)
        return version
    }

    fun version(platform: Platform, version: String): Version {
        return version(Version(platform, version, "${platform.name.lowercase()}/$version"))
    }

    fun version(platform: Platform, version: String, callback: Action<Version>): Version {
        val server = version(platform, version)
        callback.execute(server)
        return server
    }

    fun versions(platform: Platform, vararg versions: String): MultipleVersions {
        val list = mutableListOf<Version>()
        for (version in versions) {
            list.add(version(platform, version))
        }
        return MultipleVersions(list)
    }

    fun versions(platform: Platform, versions: Collection<String>): MultipleVersions {
        val list = mutableListOf<Version>()
        for (version in versions) {
            list.add(version(platform, version))
        }
        return MultipleVersions(list)
    }

    fun versions(
        platform: Platform,
        versions: Collection<String>,
        callback: Action<MultipleVersions>
    ): MultipleVersions {
        val multipleVersions = versions(platform, versions)
        callback.execute(multipleVersions)
        return multipleVersions
    }

    fun paper(version: String): Version {
        return version(Platform.PAPER, version)
    }

    fun paper(version: String, callback: Action<Version>): Version {
        return version(Platform.PAPER, version, callback)
    }

    fun paperVersions(vararg versions: String): MultipleVersions {
        return versions(Platform.PAPER, *versions)
    }

    fun paperVersions(versions: Collection<String>): MultipleVersions {
        return versions(Platform.PAPER, versions)
    }

    fun paperVersions(versions: Collection<String>, callback: Action<MultipleVersions>): MultipleVersions {
        return versions(Platform.PAPER, versions, callback)
    }

    fun pluginJar(pluginJar: Provider<RegularFile>) {
        this.pluginJar = pluginJar
    }

    @ApiStatus.Obsolete
    fun pluginJar(pluginJar: File) {
        this.pluginJar = providers.provider { RegularFile { pluginJar } }
    }

    fun testingDirectory(testingDirectory: String) {
        this.testingDirectory = testingDirectory
    }

    override fun toString(): String {
        return ("RunTestServerExtension(providers=" + this.providers
                + ", versions=" + this.versions
                + ", pluginJar=" + this.pluginJar
                + ", testingDirectory=" + this.testingDirectory + ")")
    }
}
